File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.java
class Solution {
    List<String> res = new ArrayList<>();

    private void backtrack(HashMap<Integer, String> map, String s, int i, StringBuilder sb){
        if(i==s.length()){
            res.add(sb.toString());
            return;
        }

        int digit = (int)s.charAt(i) - '0';

        for(char c: map.get(digit).toCharArray()){
            sb.append(c);
            backtrack(map, s, i+1, sb);
            sb.deleteCharAt(sb.length()-1);
        }
    }

    public List<String> letterCombinations(String digits) {
        if(digits.length()==0) return res;
        
        HashMap<Integer,String> map=new HashMap<>();
        map.put(2,"abc");
        map.put(3,"def");
        map.put(4,"ghi");
        map.put(5,"jkl");
        map.put(6,"mno");
        map.put(7,"pqrs");
        map.put(8,"tuv");
        map.put(9,"wxyz");

        backtrack(map, digits, 0, new StringBuilder()); 

        return res;
    }
}

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.py
class Solution(object):
    def backtrack(self, digitToString, digits, i, s):
        if i == len(digits):
            self.res.append(s)
            return

        digit = int(digits[i])

        for c in digitToString[digit]:
            s+=c
            self.backtrack(digitToString, digits, i+1, s)
            s = s[:-1]

        

    def letterCombinations(self, digits):
        """
        :type digits: str
        :rtype: List[str]
        """

        if len(digits) == 0:
            return []

        digitToString = {
            2 : "abc",
            3: "def",
            4: "ghi",
            5: "jkl",
            6: "mno",
            7: "pqrs",
            8: "tuv",
            9: "wxyz"
        }   

        self.res = []
        self.backtrack(digitToString, digits, 0, "")
        return self.res


        